docs/Operators.md

```diff
@@ -4943,6 +4983,163 @@ expect(node, inputs=[x], outputs=[y],
+      ) {
+        %my_local = Add(%a, %b)
+        %b_out = Sub(%a, %b)
+        %keepgoing_out = Greater(%my_local, %b_out)
+        %user_defined_vals = Add(%b, %b)
+        return %keepgoing_out, %b_out, %user_defined_vals
+      }
+  
+  *Sample equivalent C code*
+  
+      {
+        /* User-defined code (enclosing scope) */
+        int a = 3, b = 6;
+        bool keepgoing = true; // Analogous to input cond
+        /* End user-defined code */
+  
+        /* Implicitly-defined code */
+        const int max_trip_count = 10; // Analogous to input M
+        int user_defined_vals[]; // Imagine this is resizable
+        /* End implicitly-defined code */
+        for (int i=0; i < max_trip_count && keepgoing; ++i) {
+          /* User-defined code (loop body) */
+          int my_local = a + b; // Reading values in the enclosing scope is fine
+          b = a - b; // writes fine if we specify b as a loop-carried dependency
+          keepgoing = my_local > b; // keepgoing is a loop-carried dependency
+          user_defined_vals[i] = b + b;
+          /* End user-defined code */
+        }
+        // my_local = 123; // Can't do this. my_local was defined in the the body
+  
+        // These below values are live-out from the loop and therefore accessible
+        b_out; user_defined_vals; keepgoing_out;
+      }
+  
+  There are several things of note in this code snippet:
+  
+  1) Values from the enclosing scope (i.e. variable a here) are in scope and can
+     be referenced in the inputs of the loop.
+  2) Any variables which you wish to make available in the enclosing scope (i.e.
+     the variables b and keepgoing) must be declared as either loop-carried
+     dependencies (both at the op inputs and output and at the body net input and
+     output) or scan_outputs.
+  3) Values created in the body cannot be accessed in the enclosing scope.
+  
+  Note that the semantics of this op support "diagonal" or "wavefront" execution.
+  (See Step 3 here for an example:
+  https://devblogs.nvidia.com/optimizing-recurrent-neural-networks-cudnn-5/).
+  Frontends should emit multi-layer RNNs as a series of While operators (with
+  time being the inner looping dimension), with each successive layer consuming
+  the scan_outputs from the previous layer, possibly going through several
+  point-wise operators (e.g. dropout, residual connections, linear layer).
+
+#### Version
+
+This version of the operator has been available since version 1 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>body</tt> : graph (required)</dt>
+<dd>The graph run each iteration. It has 2+N inputs: (iteration_num, condition, loop carried dependencies...). It has 1+N+K outputs: (condition, loop carried dependencies..., scan_outputs...). Each scan_output is created by concatenating the value of the specified output value at the end of each iteration of the loop. It is an error if the dimensions of these values change across loop iterations.</dd>
```

Comment:
Why we have this restriction?

Reply:
Because it's invalid to concatenate non-matching tensor shapes? I think the same is the case for Loop op actually (padding on max seq size is different)

Reply:
@dzhulgakov , is it better to explain how the tensors gets concatenated (and add an example)? For example, if I have 3 iterations and each iteration generates a 7x5 tensor as the only scan_output, is the final scan_output a 3x7x5 tensor? There are multiple ways to concatenate tensors, I believe.

Reply:
it's probably more general to concatenate  by the first dimension. Then individual loop iteration can return not only slices of dim 1 (and can insert unsqeeze as needed).

So in your example each iteration would return 1x7x5 to end up with 3x7x5 output. Or each iteration can return 2x7x5, 0x7x5 and 4x7x5 making the output 6x7x5

Then we also need to add which  dimension to concat on similarly to Scan? That one actually has only input axis but hardcoded output axis to be 1.

Reply:
@dzhulgakov do we need the concatenation? Can we pass the needed data as extra parameter?

Reply:
I just talked to Rama, can we clarify the doc, is the restriction only for the scan_outputs or the scan_outputs and loop carry dependencies? We don't need restriction for the loop carry dependency, correct?